test/NameBinding/dynamic-member-lookup.swift

```diff
@@ -0,0 +1,231 @@
+  d.woof = global     // ok
+  dm.woof = global    // ok
+}
+
+
+func test_iuo(a : Gettable!, b : Settable!) {
+  global = a.wyverns
+  a.flavor = global  // expected-error {{cannot assign through dynamic lookup property: subscript is get-only}}
+  
+  global = b.flavor
+  b.universal = global // expected-error {{cannot assign through dynamic lookup property: 'b' is a 'let' constant}}
+  
+  var bm : Settable! = b
+  
+  global = bm.flavor
+  bm.universal = global
+}
+
+@dynamicMemberLookup
+struct FnTest {
+  subscript(dynamicMember member: StaticString) -> (_ a : Int)->() {
+    return { a in () }
+  }
+}
+func test_function(x : FnTest) {
+  x.phunky(12)
+}
+func test_function_iuo(x : FnTest!) {
+  x.flavor(12)
+}
+
+
+//===----------------------------------------------------------------------===//
+// Error cases
+//===----------------------------------------------------------------------===//
+
+// Subscript index must be ExpressibleByStringLiteral.
+@dynamicMemberLookup
+struct Invalid1 {
+  // expected-error @+1 {{@dynamicMemberLookup attribute requires 'Invalid1' to have a 'subscript(dynamicMember:)' member with a string index}}
+  subscript(dynamicMember member: Int) -> Int {
+    return 42
+  }
+}
+
+// Subscript may not be variadic.
+@dynamicMemberLookup
+struct Invalid2 {
+  // expected-error @+1 {{@dynamicMemberLookup attribute requires 'Invalid2' to have a 'subscript(dynamicMember:)' member with a string index}}
+  subscript(dynamicMember member: String...) -> Int {
+    return 42
+  }
+}
+
+// References to overloads are resolved just like normal subscript lookup:
+// they are either contextually disambiguated or are invalid.
+@dynamicMemberLookup
+struct Ambiguity {
+  subscript(dynamicMember member: String) -> Int {
+    return 42
+  }
+  subscript(dynamicMember member: String) -> Float {
+    return 42
+  }
+}
+
+func testAmbiguity(a : Ambiguity) {
+  let _ : Int = a.flexibility
+  let _ : Float = a.dynamism
+  _ = a.dynamism  // expected-error {{ambiguous use of 'subscript(dynamicMember:)'}}
+}
+
+
+// expected-error @+1 {{'@dynamicMemberLookup' attribute cannot be applied to this declaration}}
+@dynamicMemberLookup
+extension Int {
+  subscript(dynamicMember member: String) -> Int {
+    fatalError()
+  }
+}
+
+@dynamicMemberLookup  // expected-error {{'@dynamicMemberLookup' attribute cannot be applied to this declaration}}
+func NotAllowedOnFunc() {
+}
+
+//===----------------------------------------------------------------------===//
+// Test Existential
+//===----------------------------------------------------------------------===//
+
+@dynamicMemberLookup
+protocol PyVal {
+  subscript(dynamicMember member: StaticString) -> PyVal { get nonmutating set }
+}
+extension PyVal {
+  subscript(dynamicMember member: StaticString) -> PyVal {
+    get { fatalError() } nonmutating set {}
+  }
+}
+
+struct MyType : PyVal {
+}
+
+
+func testMutableExistential(a : PyVal, b : MyType) -> PyVal {
+  a.x.y = b
+  b.x.y = b
```

Comment:
This is hitting the narrow case where an existential is represented as a nominal type (which is why the test succeeds now); I'd expect that, e.g., `AnyObject & PyVal` would fail because that's represented as a protocol composition type.

Reply:
Fixed, thank you again for your testcase.